packages/react-compose/src/next/getSlots.tsx

```diff
@@ -0,0 +1,52 @@
+import * as React from 'react';
+import { getNativeElementProps, omit } from '@uifabric/utilities';
+import { GenericDictionary } from './types';
+import { nullRender } from './nullRender';
+
+/**
+ * Given the state and an array of slot names, will break out `slots` and `slotProps`
+ * collections.
+ *
+ * The root is always derived from the `as` prop.
+ *
+ * Slots will render as null if they are rendered as primitives with undefined children.
+ *
+ * The slotProps will always omit the `as` prop within them, and for slots that are string
+ * primitives, the props will be filtered according the the slot type. For example, if the
+ * slot is rendered `as: 'a'`, the props will be filtered for acceptable anchor props.
+ *
+ * @param state - State including slot definitions
+ * @param slotNames - Name of which props are slots
+ * @returns An object containing the `slots` map and `slotProps` map.
+ */
+export const getSlots = (state: GenericDictionary, slotNames?: string[] | undefined) => {
+  const slots: GenericDictionary = {
+    root: state.as || nullRender,
+  };
+  const slotProps: GenericDictionary = {
+    root: typeof state.as === 'string' ? getNativeElementProps(state.as, state) : omit(state, { as: 1 }),
```

Comment:
This API looks a bit weird to me:
```
 omit(state, { as: 1 })
```

Can we refactor to something like
```
omit(state, ['as'])
```

Reply:
No; this is what lodash does, and it means that every prop must call indexOf on the array, which is slot. I agree it looks weird, but it is necessary for perf, which was the whole purpose of adding omit.

Reply:
https://jsperf.com/omit-vs-rest-vs-reduce/6 see "indexOf" solution perf.

Reply:
Turns out I was wrong earlier on perf. For short lists of exclusions, indexOf seems faster than object maps. Who'd have thunk. Here's the perf results: https://jsperf.com/omit-vs-rest-vs-reduce/10

And here's the PR to fix: https://github.com/microsoft/fluentui/pull/14313

Reply:
merged and updated.